File: SpringBoot-JPA-Auditing-MySQL-Example/src/main/java/com/hoangtien2k3/controller/UserController.java
package com.hoangtien2k3.controller;

import com.hoangtien2k3.exception.ResourceNotFoundException;
import com.hoangtien2k3.model.User;
import com.hoangtien2k3.repository.UserRepository;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1")
public class UserController {

    @Autowired
    private UserRepository userRepository;

    @GetMapping("/users/{id}")
    public ResponseEntity<User> getUserById(@PathVariable(value = "id") Long userId) throws ResourceNotFoundException {
        User user = userRepository
                .findById(userId) // trả về Optional
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userId));

        return ResponseEntity.ok().body(user); // trả về body và status code OK
    }

    @GetMapping("/users")
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    @PostMapping("/users")
    public User createUser(@Valid @RequestBody User user) {
        return userRepository.save(user);
    }

    @PutMapping("/users/{id}")
    public ResponseEntity<User> updateUser(
            @PathVariable(value = "id") Long userId,
            @Valid @RequestBody User userDetails) throws ResourceNotFoundException {

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found :: " + userId));

        user.setEmailId(userDetails.getEmailId());
        user.setLastName(userDetails.getLastName());
        user.setFirstName(userDetails.getFirstName());
        user.setEmailId(userDetails.getEmailId());
        final User updatedUser = userRepository.save(user);

        return ResponseEntity.ok(updatedUser);
    }

    @DeleteMapping("/users/{id}")
    public Map<String, Boolean> deleteUser(
            @PathVariable(value = "id") Long userId) throws ResourceNotFoundException {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found :: " + userId));

        userRepository.delete(user);
        Map<String, Boolean> map = new HashMap<>();
        map.put("deleted", Boolean.TRUE);

        return map;
    }

}

File: Project-Spring-Boot/library-management-system/src/main/kotlin/com/hoangtien2k3/librarymanagementsystem/controller/UserController.kt
package com.hoangtien2k3.librarymanagementsystem.controller

import com.hoangtien2k3.librarymanagementsystem.entity.User
import com.hoangtien2k3.librarymanagementsystem.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/users")
class UserController {
    @Autowired
    lateinit var userService: UserService

    @GetMapping
    fun getAllUsers(): List<User> {
        return userService.findAll()
    }

    @PostMapping
    fun addUser(@RequestBody user: User): User {
        return userService.save(user)
    }

}